### PR TITLE
Quantize LSTM

### PR DESCRIPTION
**Pull request by levzlotnik**
Initial commit with LSTM implementation for quantization purposes

### CODE REVIEW

File: distiller/modules/concat.py [hunk 1]
```diff
@@ -25,3 +25,32 @@ def __init__(self, dim=0):
 
     def forward(self, *seq):
         return torch.cat(seq, dim=self.dim)
+
+
+class Chunk(nn.Module):
```
Comment (nzmora) [inline distiller/modules/concat.py]: The name of the file is concat.py but you added several somewhat related layers into one file.  You should either change the file name, or place each PyTorch module in its own file (preferred).

File: examples/word_language_model/model.py [hunk 2]
```diff
@@ -58,3 +60,133 @@ def init_hidden(self, bsz):
                     weight.new_zeros(self.nlayers, bsz, self.nhid))
         else:
             return weight.new_zeros(self.nlayers, bsz, self.nhid)
+
+
+class LSTMCell(nn.Module):
```
Comment (nzmora) [inline examples/word_language_model/model.py]: this should go in distiller/modules/lstm.py

File: examples/word_language_model/model.py [hunk 3]
```diff
@@ -58,3 +60,133 @@ def init_hidden(self, bsz):
...
+        super(LSTMCell, self).__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+
+        # Treat f,i,o,c_ex as one single object:
+        self.fc_gate_x = nn.Linear(input_size, hidden_size * 4)
+        self.fc_gate_h = nn.Linear(hidden_size, hidden_size * 4)
+        self.eltwiseadd_gate = EltwiseAdd()
+        # Split the object into f,i,o,c_ex:
+        self.chunk_gate = Chunk(4, dim=1)
```
Comment (nzmora) [inline examples/word_language_model/model.py]: I don't think you should use a module for chunking.  You can do the chunking in the ```forward``` pass as done here: https://github.com/NervanaSystems/distiller/blob/642996cb0df25c1b69e44f66bb83f31aa1cd410c/distiller/models/cifar10/simplenet_cifar.py#L41 to create a view of a tensor.

File: examples/word_language_model/model.py [hunk 4]
```diff
@@ -58,3 +60,133 @@ def init_hidden(self, bsz):
...
+        c_0 = torch.zeros(batch_size, self.hidden_size).to(device)
+        return h_0, c_0
+
+    def init_weights(self):
+        initrange = 0.1
+        self.fc_gate_x.weight.data.uniform_(-initrange, initrange)
+        self.fc_gate_h.weight.data.uniform_(-initrange, initrange)
+
+
+class LSTM(nn.Module):
```
Comment (nzmora) [inline examples/word_language_model/model.py]: this should go in distiller/modules/lstm.py

File: distiller/modules/concat.py [hunk 5]
```diff
@@ -25,3 +25,32 @@ def __init__(self, dim=0):
...
+    def __init__(self, split_size_or_sections, dim=0):
+        super(Split, self).__init__()
+        self.split_size_or_sections = split_size_or_sections
+        self.dim = dim
+
+    def forward(self, tensor):
+        return torch.split(tensor, self.split_size_or_sections, dim=self.dim)
+
+
+class Stack(nn.Module):
```
Comment (nzmora) [inline distiller/modules/concat.py]: You don't use this class, so you should delete this code.

File: distiller/modules/concat.py [hunk 6]
```diff
@@ -25,3 +25,32 @@ def __init__(self, dim=0):
...
+    def __init__(self, chunks, dim=0):
+        super(Chunk, self).__init__()
+        self.chunks = chunks
+        self.dim = dim
+
+    def forward(self, tensor):
+        return tensor.chunk(self.chunks, dim=self.dim)
+
+
+class Split(nn.Module):
```
Comment (nzmora) [inline distiller/modules/concat.py]: You don't use this class, so you should delete this code.

File: examples/word_language_model/model.py (hunk 3 above)
Comment (nzmora) [reply examples/word_language_model/model.py]: Hi Lev, I didn’t explain myself well.  I mean to say that you can use ```torch.chunk``` in the ```forward``` function, and that you don't need to create a torch module just to do this.  The ```view``` example above, was meant to show how tensors can be manipulated in the forward pass, without requiring a module.

File: examples/word_language_model/model.py (hunk 2 above)
Comment (levzlotnik) [reply examples/word_language_model/model.py]: created new file distiller/modules/rnn.py

**Comment by levzlotnik** [comment]
Note - in the last commit I renamed `slice_and_join.py` into `grouping.py` (which was originally `concat.py`).

File: examples/word_language_model/model.py [hunk 7]
```diff
@@ -58,3 +65,34 @@ def init_hidden(self, bsz):
                     weight.new_zeros(self.nlayers, bsz, self.nhid))
         else:
             return weight.new_zeros(self.nlayers, bsz, self.nhid)
+
+
+class WordLangModel(nn.Module):
```
Comment (nzmora) [inline examples/word_language_model/model.py]: What's this class?  Why is it needed and where is it used?

File: distiller/modules/grouping.py [hunk 8]
```diff
@@ -25,3 +25,32 @@ def __init__(self, dim=0):
 
     def forward(self, *seq):
         return torch.cat(seq, dim=self.dim)
+
+
+class Chunk(nn.Module):
```
Comment (nzmora) [inline distiller/modules/grouping.py]: You haven't provided an explanation for why we need these new modules that you don't use.
Code that is DOA can only become a burden (maintenance; confusion).

**Comment by Coulson1026** [comment]
> Initial commit with LSTM implementation for quantization purposes

Excuse me, does the code you submitted support BiLSTM?

**Comment by levzlotnik** [comment]
@Coulson1026 
> > Initial commit with LSTM implementation for quantization purposes
> 
> Excuse me, does the code you submitted support BiLSTM?
Hi, the code doesn't support Bidirectional LSTM yet, we added this feature to try out our quantization methods in an LSTM workload context. When we merge this PR, feel free to add a bidirectional flag similar to `torch.nn.LSTM`. At the moment we prioritize on quantizing the LSTM itself.

**Comment by Coulson1026** [comment]
> @Coulson1026
> 
> > > Initial commit with LSTM implementation for quantization purposes
> > 
> > 
> > Excuse me, does the code you submitted support BiLSTM?
> 
> Hi, the code doesn't support Bidirectional LSTM yet, we added this feature to try out our quantization methods in an LSTM workload context. When we merge this PR, feel free to add a bidirectional flag similar to `torch.nn.LSTM`. At the moment we prioritize on quantizing the LSTM itself.

Thanks, do you quantize the intermediate activations within the cells? I didn't find any code related to it.

**Comment by levzlotnik** [comment]
@Coulson1026 
Actually, when using any `distiller.quantization.Quantizer`, 
and in particular `distiller.quantization.PostTrainLinearQuantizer` - the activations are quantized (or not, depending on the configuration you provide) regardless of their position in the graph. 
In short - yes, the intermediate activations can be quantized, see reference guide for configuring `Quantizer`s [here](https://nervanasystems.github.io/distiller/schedule.html#post-training-quantization)

**Comment by levzlotnik** [comment]
@Coulson1026 
Soon I will push a jupyter notebook for quantizing an LSTM.

**Comment by Coulson1026** [comment]
Thank you for your help

**Comment by levzlotnik** [comment]
Hi @Coulson1026 ,
We added BiLSTM, feel free to try it out

**Comment by Coulson1026** [comment]
@levzlotnik OK，thanks

**Comment by guyjacob** [comment]
The latest refactoring looks good - thanks!
A few hopefully-final requests:
1. Type 1 - not supporting for now is perfectly fine, but please raise the `NotImplementedError` at `__init__` time. No need to wait until forward is called.
2. Weights initialization - you're initializing uniformly between [-0.1, 0.1]. The PyTorch implementation intializes uniformly between [-K, K], where `K = sqrt(1 / hidden_size)`. I realize we're not focused on training right now, but let's maintain the consistency with the PyTorch implementation as done so far.
3. The three functions that handle re-formatting/packaging (`_reformat_hidden`, `_reformat_hidden_bidirectional`, `_repackage_bidirectional_input_h`) - can you please add a comment in each explaining what it's converting to/from and why. (BTW - what's the difference between reformat and repackage? :) )

**Pull request merged**